ksp_plugin_adapter/orbit_analyser.cs

```diff
@@ -0,0 +1,427 @@
+      return false;
+    }
+    string days = match.Groups[1].Value;
+    string hours = match.Groups[2].Value;
+    string minutes = match.Groups[3].Value;
+    if (!int.TryParse(days, out int d) ||
+        !int.TryParse(hours, out int h) ||
+        !int.TryParse(minutes, out int min)) {
+      return false;
+    }
+    value = (TimeSpan.FromDays((double)d / (GameSettings.KERBIN_TIME ? 4 : 1)) +
+             TimeSpan.FromHours(h) +
+             TimeSpan.FromMinutes(min)).TotalSeconds;
+    return true;
+  }
+}
+
+internal class OrbitAnalyser : SupervisedWindowRenderer {
+  public OrbitAnalyser(PrincipiaPluginAdapter adapter)
+      : base(adapter, UnityEngine.GUILayout.MinWidth(0)) {
+    adapter_ = adapter;
+  }
+
+  public void RenderButton() {
+    if (UnityEngine.GUILayout.Button("Orbit analysis...")) {
+      Toggle();
+    }
+  }
+
+  protected override string Title => "Orbit analysis";
+
+  protected override void RenderWindow(int window_id) {
+    using (new UnityEngine.GUILayout.VerticalScope(GUILayoutWidth(8))) {
+      Vessel vessel = FlightGlobals.ActiveVessel;
+      if (plugin == IntPtr.Zero || vessel == null ||
+          !plugin.HasVessel(vessel.id.ToString())) {
+        Hide();
+        return;
+      }
+      CelestialBody primary =
+          adapter_.plotting_frame_selector_.selected_celestial;
+
+      mission_duration_.Render(enabled : true);
+      var multiline_style = Style.Multiline(UnityEngine.GUI.skin.label);
+      float two_lines = multiline_style.CalcHeight(
+          new UnityEngine.GUIContent("1\n2"), Width(1));
+      float five_lines = multiline_style.CalcHeight(
+          new UnityEngine.GUIContent("1\n2\n3\n4\n5"), Width(1));
+      UnityEngine.GUILayout.Label(
+          $@"Analysing orbit of {vessel.vesselName} with respect to {
+            primary.NameWithArticle()}...",
+          multiline_style,
+          UnityEngine.GUILayout.Height(two_lines));
+
+      OrbitAnalysis analysis = plugin.VesselRefreshAnalysis(
+          vessel.id.ToString(),
+          primary.flightGlobalsIndex,
+          mission_duration_.value,
+          autodetect_recurrence_ ? null : (int?)revolutions_per_cycle_,
+          autodetect_recurrence_ ? null : (int?)days_per_cycle_,
+          ground_track_revolution_);
+      if (autodetect_recurrence_ &&
+          analysis.recurrence.number_of_revolutions != 0 &&
+          analysis.recurrence.cto != 0) {
+        revolutions_per_cycle_ = analysis.recurrence.number_of_revolutions;
+        days_per_cycle_ = analysis.recurrence.cto;
+      }
+      UnityEngine.GUILayout.HorizontalScrollbar(
+          value      : 0,
+          size       : (float)analysis.progress_of_next_analysis,
+          leftValue  : 0,
+          rightValue : 1);
+
+      OrbitalElements? elements = null;
+      OrbitRecurrence? recurrence = null;
+      OrbitGroundTrack? ground_track = null;
+      double mission_duration = analysis.mission_duration;
+      if (analysis.elements_has_value) {
```

Comment:
Create a bug for supporting optional fields in the auxiliary interface messages.

Reply:
#2339.